src/structures/CommandInteraction.js

```diff
@@ -0,0 +1,195 @@
+    this.deferred = true;
+  }
+
+  /**
+   * Options for a reply to an interaction.
+   * @extends MessageOptions
+   * @property {boolean} [ephemeral] Whether the reply should be ephemeral
+   * @typedef {Object} InteractionReplyOptions
+   */
+
+  /**
+   * Creates a reply to this interaction.
+   * @param {string|APIMessage|MessageAdditions} content The content for the reply
+   * @param {InteractionReplyOptions} [options] Additional options for the reply
+   * @returns {Promise<void>}
+   */
+  async reply(content, options) {
+    const apiMessage = content instanceof APIMessage ? content : APIMessage.create(this, content, options);
+    const { data, files } = await apiMessage.resolveData().resolveFiles();
+
+    await this.client.api.interactions(this.id, this.token).callback.post({
+      data: {
+        type: InteractionResponseTypes.CHANNEL_MESSAGE_WITH_SOURCE,
+        data,
+      },
+      files,
+    });
+    this.replied = true;
+  }
+
+  /**
+   * Fetches the initial reply to this interaction.
+   * @returns {Promise<?Message>}
+   */
+  async fetchReply() {
+    const applicationID = this.client.application.id;
+    const raw = await this.client.api.webhooks(applicationID, this.token).messages('@original').patch({ data: {} });
+    return this.channel?.messages.add(raw) ?? null;
+  }
+
+  /**
+   * Edits the initial reply to this interaction.
+   * @param {string|APIMessage|MessageEmbed|MessageEmbed[]} content The new content for the message
+   * @param {WebhookEditMessageOptions} [options] The options to provide
+   * @returns {Promise<?Message>}
+   */
+  async editReply(content, options) {
+    const raw = await this.webhook.editMessage('@original', content, options);
+    return this.channel?.messages.add(raw) ?? null;
+  }
+
+  /**
+   * Deletes the initial reply to this interaction.
+   * @returns {Promise<void>}
+   */
+  async deleteReply() {
+    await this.webhook.deleteMessage('@original');
+  }
+
+  /**
+   * Represents an option of a received command interaction.
+   * @property {string} name The name of the option
+   * @property {ApplicationCommandOptionType} type The type of the option
+   * @property {string|number|boolean} [value] The value of the option
+   * @property {CommandInteractionOption[]} [options] Additional options if this option is a subcommand (group)
+   * @property {User} [user] The resolved user
+   * @property {GuildMember} [member] The resolved member
+   * @property {GuildChannel} [channel] The resolved channel
+   * @property {Role} [role] The resolved role
+   * @typedef {Object} CommandInteractionOption
+   */
+
+  /**
+   * Transforms an option received from the API.
+   * @param {Object} option The received option
+   * @param {Object} resolved The resolved interaction data
+   * @returns {CommandInteractionOption}
+   * @private
+   */
+  transformOption(option, resolved) {
+    const result = {
+      name: option.name,
+      type: ApplicationCommandOptionTypes[option.type],
```

Comment:
For the sake of DX, it would make sense to add more flexible validation to the option type. For example, setting the `type` to a lowercase string of `string` causes the field to be unset, which isn't valid:
![code](https://user-images.githubusercontent.com/442980/114205122-97845480-991f-11eb-8f93-1a6e1214cb88.png)

That could be solved easily by adding `option.type.toUpperCase()` here.

I'd also expect this to allow me to use the ApplicationCommandOptionType values directly.

```suggestion
      type: (typeof option.type === 'number')
        ? option.type
        : ApplicationCommandOptionTypes[option.type.toUpperCase()],
```

Reply:
I implemented the ability to pass numbers as types, however I think that if you pass a string, it should still be capitalized (see also e.g. #4766)